### PR TITLE
NormalizedGINI, FairLoss

### PR DESCRIPTION
**Pull request by Antoshqa**
Implemented NormalizedGini and FairLoss

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en.

### CODE REVIEW

File: catboost/libs/metrics/metric.cpp [hunk 1]
```diff
@@ -2086,6 +2086,122 @@ void TAverageGain::GetBestValue(EMetricBestValue* valueType, float*) const {
...
+    Y_ASSERT((approx.size() > 1) == IsMultiClass);
+    const auto& approxVec = approx.ysize() == 1 ? approx.front() : approx[PositiveClass];
+    Y_ASSERT(approxVec.size() == target.size());
+    const auto& weight = UseWeights ? weightIn : TVector<float>{};
+
+    TVector<double> approxCopy(approxVec.begin() + begin, approxVec.begin() + end);
+    TVector<double> targetCopy(target.begin() + begin, target.begin() + end);
+
+    if (!IsMultiClass) {
+        for (ui32 i = 0; i < targetCopy.size(); ++i) {
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/libs/metrics/metric.cpp]: Please use range-based for.

File: catboost/libs/metrics/metric.cpp [hunk 2]
```diff
@@ -2086,6 +2086,122 @@ void TAverageGain::GetBestValue(EMetricBestValue* valueType, float*) const {
...
+            targetCopy[i] = targetCopy[i] > Border;
+        }
+    }
+
+    if (approx.ysize() > 1) {
+        int positiveClass = PositiveClass;
+        ForEach(targetCopy.begin(), targetCopy.end(), [positiveClass](double& x) {
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/libs/metrics/metric.cpp]: Please use range-based for and remove local positiveClass.

File: catboost/libs/metrics/metric.cpp [hunk 3]
```diff
@@ -2086,6 +2086,122 @@ void TAverageGain::GetBestValue(EMetricBestValue* valueType, float*) const {
...
+        int begin,
+        int end
+) const {
+    CB_ENSURE(approx.size() == 1, "Metric FairLoss supports only single-dimensional data");
+
+    const auto& approxVec = approx.front();
+    Y_ASSERT(approxVec.size() == target.size());
+
+    TMetricHolder error(2);
+    for (int k = begin; k < end; ++k) {
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/libs/metrics/metric.cpp]: Please use meaningful identifiers instead of k, w and x, e.g. idx, documentWeight, targetMismatch.

File: catboost/libs/metrics/metric.h [hunk 4]
```diff
@@ -1077,6 +1077,49 @@ class TAverageGain : public TAdditiveMetric<TAverageGain> {
...
+            const TVector<float>& weight,
+            const TVector<TQueryInfo>& queriesInfo,
+            int begin,
+            int end
+    ) const;
+    virtual TString GetDescription() const override;
+    virtual double GetFinalError(const TMetricHolder& error) const override;
+    virtual void GetBestValue(EMetricBestValue* valueType, float* bestValue) const override;
+private:
+    double c;
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/libs/metrics/metric.h]: Please start all class member names with a captial letter.

File: catboost/libs/metrics/ut/fairloss_ut.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                TVector<float> weight{1, 1, 1};
+                NPar::TLocalExecutor executor;
+
+                TFairLossMetric metric(5);
+                TMetricHolder score = metric.Eval(approx, target, weight, {}, 0, target.size(), executor);
+
+                UNIT_ASSERT_DOUBLES_EQUAL(metric.GetFinalError(score), 0.441961, 1e-5);
+            }
+        }
+}
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/libs/metrics/ut/fairloss_ut.cpp]: Please add the empty line at the end of each file.

File: catboost/libs/metrics/ut/fairloss_ut.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
+#include <library/unittest/registar.h>
```
Comment (annaveronika) [inline catboost/libs/metrics/ut/fairloss_ut.cpp]: Please add to your pr two python files with a stress test for every metric.
Stress test must compare results of your implementation with some other existing implementation, for example, one from sklearn.metrics or some other one.

To use your metric from python you can use eval_metric function.

Please provide the output of your stress test in the comment.

File: catboost/libs/metrics/metric.h [hunk 7]
```diff
@@ -1102,6 +1102,24 @@ struct TNormalizedGINI: public TNonAdditiveMetric {
     double Border = GetDefaultClassificationBorder();
 };
 
+//FairLoss metric
+struct TFairLossMetric: public TAdditiveMetric<TFairLossMetric> {
+    explicit TFairLossMetric(double c = 2);
```
Comment (annaveronika) [inline catboost/libs/metrics/metric.h]: please use a better name for the parameter

File: catboost/libs/metrics/metric.h [hunk 8]
```diff
@@ -1102,6 +1102,24 @@ struct TNormalizedGINI: public TNonAdditiveMetric {
...
+    TMetricHolder EvalSingleThread(
```
Comment (annaveronika) [inline catboost/libs/metrics/metric.h]: Please add md file that explains changes that will be added to documentation here:
https://tech.yandex.com/catboost/doc/dg/concepts/loss-functions-docpage/

File: catboost/libs/metrics/ut/fairloss_ut.cpp (hunk 6 above)
Comment (Antoshqa) [reply catboost/libs/metrics/ut/fairloss_ut.cpp]: Should I add this stress tests to somewhere in the repository or is it enough to run it with my local build and just add them here as a file, providing result in comment?

File: catboost/libs/metrics/metric.h (hunk 8 above)
Comment (Evgueni-Petrov-aka-espetrov) [reply catboost/libs/metrics/metric.h]: Anton, I can help with English in the documentation :)

**Comment by annaveronika** [comment]
@Antoshqa Please update the repo with new commits from master and merge the changes.

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@Evgueni-Petrov-aka-espetrov, internal review request created: 650109

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
In file included from $(SOURCE_ROOT)/catboost/cuda/cpu_compatibility_helpers/model_converter.cpp:3:
In file included from $(SOURCE_ROOT)/catboost/libs/algo/helpers.h:3:
In file included from $(SOURCE_ROOT)/catboost/libs/algo/learn_context.h:15:
In file included from $(SOURCE_ROOT)/catboost/libs/loggers/catboost_logger_helpers.h:5:
$(SOURCE_ROOT)/catboost/libs/metrics/metric.h:302:27: error: 'Eval' marked 'override' but does not override any member functions
    virtual TMetricHolder Eval(
                          ^
$(SOURCE_ROOT)/catboost/libs/metrics/metric.h:302:27: error: 'TNormalizedGINIMetric::Eval' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
$(SOURCE_ROOT)/catboost/libs/metrics/metric.h:98:27: note: hidden overloaded virtual function 'IMetric::Eval' declared here: type mismatch at 2nd parameter ('TConstArrayRef<float>' (aka 'TArrayRef<const float>') vs 'const TVector<float> &')
    virtual TMetricHolder Eval(
                          ^

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Could you please fix the build error above?

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@Evgueni-Petrov-aka-espetrov, internal review request created: 652073

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@Evgueni-Petrov-aka-espetrov, internal review request created: 652109

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@Evgueni-Petrov-aka-espetrov, internal review request created: 652141

**Comment by arcadia-devtools** [comment]
@Evgueni-Petrov-aka-espetrov, internal review request created: 652273

File: catboost/libs/metrics/metric.cpp [hunk 9]
```diff
@@ -3166,6 +3166,122 @@ void TAverageGain::GetBestValue(EMetricBestValue* valueType, float*) const {
...
+) const {
+    Y_ASSERT((approx.size() > 1) == IsMultiClass);
+    const auto& approxVec = approx.ysize() == 1 ? approx.front() : approx[PositiveClass];
+    Y_ASSERT(approxVec.size() == target.size());
+
+    TVector<double> targetCopy(target.begin() + begin, target.begin() + end);
+
+    if (!IsMultiClass) {
+        for (ui32 target : targetCopy) {
+            target = target > Border;
```
Comment (yazevnul) [inline catboost/libs/metrics/metric.cpp]: you are modifying local copy, this is noop and probably a bug that need a test case

File: catboost/libs/metrics/metric.cpp [hunk 10]
```diff
@@ -3166,6 +3166,122 @@ void TAverageGain::GetBestValue(EMetricBestValue* valueType, float*) const {
...
+        }
+    }
+
+    if (approx.ysize() > 1) {
+        for (ui32 target : targetCopy) {
+            target = (target == static_cast<double>(PositiveClass));
```
Comment (yazevnul) [inline catboost/libs/metrics/metric.cpp]: this is the same bug as above

File: catboost/libs/metrics/metric.cpp [hunk 11]
```diff
@@ -3166,6 +3166,122 @@ void TAverageGain::GetBestValue(EMetricBestValue* valueType, float*) const {
...
+    error.Stats[0] = normalizedGini;
+    error.Stats[1] = 1.0;
+    return error;
+}
+
+TString TNormalizedGiniMetric::GetDescription() const {
+    if (IsMultiClass) {
+        const TMetricParam<int> positiveClass("class", PositiveClass, /*userDefined*/true);
+        return BuildDescription(ELossFunction::NormalizedGini, UseWeights, positiveClass);
+    } else {
```
Comment (yazevnul) [inline catboost/libs/metrics/metric.cpp]: imho, this can be simplified:
```suggestion
     } 
     
     return BuildDescription(ELossFunction::NormalizedGini, UseWeights, MakeBorderParam(Border));
```

File: catboost/libs/metrics/metric.cpp [hunk 12]
```diff
@@ -3166,6 +3166,122 @@ void TAverageGain::GetBestValue(EMetricBestValue* valueType, float*) const {
...
+
+TMetricHolder TFairLossMetric::EvalSingleThread(
+        const TVector<TVector<double>>& approx,
+        TConstArrayRef<float> target,
+        TConstArrayRef<float> weight,
+        TConstArrayRef<TQueryInfo> /*queriesInfo*/,
+        int begin,
+        int end
+) const {
+    CB_ENSURE(approx.size() == 1, "Metric FairLoss supports only single-dimensional data");
```
Comment (yazevnul) [inline catboost/libs/metrics/metric.cpp]:
```suggestion
     CB_ENSURE(approx.size() == 1, "Metric FairLoss supports only single-dimensional data; " << LabeledOutput(approx.size()));
```

File: catboost/libs/metrics/metric.h [hunk 13]
```diff
@@ -290,6 +290,49 @@ THolder<IMetric> MakeUserDefinedQuerywiseMetric(const TMap<TString, TString>& pa
 THolder<IMetric> MakeAverageGainMetric(float topSize);
 
 
+//Normalized Gini metric
+struct TNormalizedGiniMetric: public TNonAdditiveMetric {
```
Comment (yazevnul) [inline catboost/libs/metrics/metric.h]: This struct and `TNormalizedGiniMetric` must be moved to `.cpp` file and put into an anonymous namespace.

End user should not have access to solid class.

File: catboost/libs/metrics/metric.h [hunk 14]
```diff
@@ -290,6 +290,49 @@ THolder<IMetric> MakeUserDefinedQuerywiseMetric(const TMap<TString, TString>& pa
...
+    virtual TMetricHolder Eval(
+            const TVector<TVector<double>>& approx,
+            TConstArrayRef<float> target,
+            TConstArrayRef<float> weight,
+            TConstArrayRef<TQueryInfo> queriesInfo,
+            int begin,
+            int end,
+            NPar::TLocalExecutor& executor) const override;
+    virtual TString GetDescription() const override;
+    virtual void GetBestValue(EMetricBestValue* valueType, float* bestValue) const override;
```
Comment (yazevnul) [inline catboost/libs/metrics/metric.h]: it's better to omit `virtual` as of `C++11` when you override virtual method

File: catboost/libs/metrics/stress-test/fairloss_stress_text.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
+import numpy as np
```
Comment (yazevnul) [inline catboost/libs/metrics/stress-test/fairloss_stress_text.py]: all these test are not going be run in our CI

you should write a test using `pytest` framework, see `util/tests/cython`, `catboost/pytest` or `library/python/strings` for examples

File: catboost/libs/metrics/stress-test/fairloss_stress_text.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    c = 2
+    res = np.mean(c**2 * ((x / c) - np.log((x/c) + 1)))
+    return res
+
+list_sizes = [10**exp for exp in range(1, 7)]
+
+for size in list_sizes:
+    print("Evaluating for dataset size =", size)
+    for k in range(100):
+        actual = np.random.randint(0, 2, size)
```
Comment (yazevnul) [inline catboost/libs/metrics/stress-test/fairloss_stress_text.py]: you test is not deterministic, you should use `np.random.RandomState` with fixed `seed` for random number generation.

File: catboost/libs/metrics/ut/fairloss_ut.cpp [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+#include <catboost/libs/metrics/metric.h>
+#include <catboost/libs/metrics/metric_holder.h>
+
+Y_UNIT_TEST_SUITE(FairLossMetricTest) {
+        Y_UNIT_TEST(FairLossTest) {
+            {
+                TVector <TVector<double>> approx{{1.0, 2, 3}};
```
Comment (yazevnul) [inline catboost/libs/metrics/ut/fairloss_ut.cpp]: unfortunately these unit tests need to be rewritten, you should create metric evaluator through `CreateMetricsFromDescription` factory function; seed `*DCG` metric tests for example (`catboost/libs/metrics/ut/dcg_ut.cpp`)

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Many thanks for the PR!
Do plan to continue working on it?

**Comment by annaveronika** [comment]
There is no activity for more than a month. Closing the pr. That's a pity that we didn't finish it.
